### ISSUE TITLE
2.3.1 pypi broken

### ISSUE THREAD

**Issue by brianmay** [aio-libs/aiohttp#2347]
The 2.3.1 upload to pypi only includes files usable on Windows.

<img width="968" alt="screenshot 2017-10-19 15 39 02" src="https://user-images.githubusercontent.com/112729/31754472-a6cc290a-b4e3-11e7-99c7-daf007039a07.png">

**Comment by fafhrd91** [comment]
Travis build is still running, it will eventually upload all required packages

**Comment by brianmay** [comment]
Ok. I think I understand now. Thanks.

The problem I have (and I have no idea who to blame or what/where the solution is) is that https://pyup.io/ sees the update as soon as the first upload occurs. Then Travis fails the PR because Travis can't actually download the updated version yet.

**Comment by brianmay** [comment]
Example here: https://travis-ci.org/brianmay/robotica/builds/289776806

**Comment by brianmay** [comment]
For future reference, the Travis build for aiohttp is here: https://travis-ci.org/aio-libs/aiohttp/builds/289766722

**Comment by edmorley** [comment]
Hi! 

I presume the issue here is that the PyPI builds for aiohttp are generated via two different means - TravisCI for Linux/OS X and CircleCI for OS X, and that the latter doesn't run any tests so occurs almost instantly, whereas the former doesn't upload until 40 minutes later, after the tests have run.

It seems like shortening the test duration for Travis (eg by running the linux/OS X tests in parallel, rather than in series, and running on the faster fully-virtualised GCE infra) and also adding tests for CircleCI would at least bring the 2mins vs 40 mins divide closer. Failing that, perhaps the CircleCI publish step should just include a `sleep` 😆 

If the gap were more like <5 minutes, then it would seem reasonable for pyup to always just pause for a few minutes before opening the PR. (This would also have the added benefit of allowing time for the tagged GitHub commits to be pushed for people who use a manual workflow, ensuring the pyup generated changelog included the most recent release.)

Fwiw I've never seen this problem with other packages, so there must be a way to make this work :-)

CC @jayfk from pyup.io.

**Comment by brianmay** [comment]
The time was much greater then 40 minutes. The PR https://github.com/brianmay/robotica/pull/40 was first created over 8 hours, it was only recently - e.g. in the last hour that all the files appeared.

**Comment by asvetlov** [comment]
@edmorley your investigation is almost right.
appveyor uploads binary Windows wheels much faster than travis deploys tarball.
Now the build is fixed, we are working on preventing issues like this in future.

**Comment by jayfk** [comment]
Also, I could add a delay for aiohttp on pyup.io. Getting CI builds in sync is not that easy and since aiohttp is a widely used package it might make sense to do it.

**Comment by asvetlov** [comment]
@jayfk let's postpone adding a delay.
I'll try to fix the problem in aiohttp itself (multidict and yarl are also affected).
But thank you for quick response.
It's good to know a direct contact in pyup bot team ;)

**Comment by edmorley** [comment]
Sounds good - many thanks :-)

In case it helps, the faster Travis GCE infra I mentioned before is used via `sudo: required`. It's possible to mix and match (eg use the slower containers that have faster boot time for the short lived lint jobs, and then the slow-to-boot but faster CPU GCE infra for everything else). See the comparison table here:
https://docs.travis-ci.com/user/reference/overview/#Virtualization-environments

**Comment by asvetlov** [comment]
Thanks.

@webknjaz maybe you are interesting in the mentioned article

**Comment by webknjaz** [comment]
@edmorley I know about this virtualization difference, but the main issue is that Travis' OS X machines are:
1) damn slow
2) restricted to run more than 2 VMs per open-source repo simultaneously

I've been thinking about adding Circle CI into play, but this (1) requires certain hacks and (2) I wouldn't use them for building wheels, because of missynchronization with Travis and AppVeyor would cause having published OS X wheels before even all other tests in other CIs are finished.
I would like to have some lock for that so build/publish activities on all CIs would wait until all the test activities in all possible envs are complete.

**Comment by webknjaz** [comment]
Ref https://blog.travis-ci.com/2017-09-22-macos-update:

> After much painstaking deliberation, we have decided to limit the concurrent macOS builds to a **maximum of two out of the five** offered. If you are using Linux only on .org, you will still have access to all five concurrent builds, but **if you are running both, only two of the five concurrent** builds can be used for macOS at any one time.

**Comment by edmorley** [comment]
Yeah that definitely isn't helping, but at the moment this project's Travis config uses the "build stages" feature in a way that means that none of the OS X jobs start until all of the Linux jobs have completed. Running them in parallel (in the same stage) would reduce the impact of the OS X backlogs.

> I would like to have some lock for that so build/publish activities on all CIs would wait until all the test activities in all possible envs are complete.

Perhaps the way to do this is to stop Circle CI triggering on-commit, but have the Travis deploy step trigger manual builds using the API? (https://circleci.com/docs/api/v1-reference/#new-build)

**Comment by edmorley** [comment]
I would also remove the python nightly and 3.N-dev permutations from the OS X run, and just leave those to Linux.

**Comment by webknjaz** [comment]
@edmorley yeah, I was thinking about it. Still, this requires some time for implementation.

So yes, I'm going to experiment with:
- [ ] combining OS X jobs with GNU/Linux tests (currently, there's an `if: ` pre-processor check, which prevents Travis from running OS X at stage level, so we'll need to move that to job definitions)
- [ ] playing with API for triggering other CI builds
- [ ] hacking Circle CI
- [ ] dropping unstable pythons from OS X matrix

I'll do this with multidict first, as it's our staging area currently.


BTW, publishing OS X wheels is stuck currently because of Travis's scheduler I guess:
> #7624.19 queued
> Waiting to start

**Comment by brianmay** [comment]
Thanks everyone!

**Comment by asvetlov** [comment]
Just note: I've added tarball upload to appveyor.
Now tarball appears first on PyPI.
Binary wheels are uploaded when ready.
I think we need nothing to do more, at least I don't want invest my time into the problem.

**Comment by edmorley** [comment]
Thank you for everyone's time spent looking at this so far.

Just to repeat what I said in aio-libs/multidict#182, the sdist appveyor upload definitely helps, though it will still cause breakage for tooling that pins the hashes in requirements files, since once the manylinux wheels are uploaded, pip will prefer the wheels over the sdist, and that new hash won't be present in the requirements file.

Really what's needed here is a PyPI RSS feed for "new file uploaded" (which would allow tools to update the pinned hashes each time), however as far as I know this doesn't exist (the existing feed only detects the first time a new version is seen in any form).

As such, the next best solution is for tools like pyup.io to just add a delay (eg 1 hour) so all wheels have been generated. However in this case it would mean a 12-24 hour delay, given the unpredictability of the OS X backlogs, which are holding up the manylinux jobs. (Shame the OS X parts can't be split out to a completely separate Travis config.)

**Comment by edmorley** [comment]
> Really what's needed here is a PyPI RSS feed for "new file uploaded"

I've filed pypa/warehouse#2551 for this.

**Comment by asvetlov** [comment]
As I understand you are asking for uploading linux wheels together with tarball at first.
It fixes installs for Linux but breaks currently working Windows in the same way.
I don't think the request worth to be implemented.
The proper solution for your needs is new feed as you wrote.
PyPI is Open Source project, please contribute the feature into it instead of enforcing on aiohttp/multidict/whatever deployment change in counternatural way :)

**Comment by asvetlov** [comment]
@edmorley thanks for issue in warehouse.
Found your answer just after sending my last comment.

**Comment by edmorley** [comment]
pyup.io has recently added a 2 hour delay between a new aiohttp/multidict release and opening the corresponding requirements.txt update PRs, however this turns out to not be enough since for the 2.3.5 release there was a 7.5 hour delay between the sdist and manylinux1 uploads (9 hours between sdist and the last OS X wheel). See:
https://github.com/pyupio/pyup/issues/250#issuecomment-348192691

**Comment by edmorley** [comment]
I don't disagree, though there aren't any good workarounds for downstream consumers at the moment (pypa/warehouse#2551 isn't going to happen overnight), and so if there's anything at all that can be done to help improve the situation on aiohttp/multidict's side, it would be really helpful. I hadn't seen #2574, that sounds great.

**Comment by webknjaz** [comment]
Thanks for the info, I'll keep this in mind as well. Hopefully I'll be able to find some compromise solution.
